### ISSUE TITLE
Get detector talking to epics

### ISSUE THREAD

**Issue by mrakitin** [NSLS-II-ESM/profile_collection#37]

**Comment by mrakitin** [comment]
➤ John Sinsheimer commented:

We are presently waiting on the vendor to provide installation instructions and/or drivers for our new Windows 10 installation, which was required for the new software.

**Comment by mrakitin** [comment]
➤ Dan Allan commented:

John -- Still waiting on installation instructions? Just checking that DAMA is not the blocking issue.

**Comment by mrakitin** [comment]
➤ John Sinsheimer commented:

I'm very close to communicating with the instrument, but just tracked down the bug I'm hitting today, and emailing about it momentarily.

**Comment by mrakitin** [comment]
➤ John Sinsheimer commented:

Another update on this - after wasting a lot of time and sending many more emails, we've found that the new software we've been trying to install is not compatible with our hardware.  But, it sounds like the incompatible interface is upgradable, and they will send us a new interface.  No ETA yet however on when this will happen.

**Comment by mrakitin** [comment]
➤ Dan Allan commented:

Thanks for the update.

**Comment by mrakitin** [comment]
➤ John Sinsheimer commented:

Another update - we received a new ethernet based interface which should allow us to use the new software.  Switching over the hardware is going to require some work and testing, but now users are using the analyzer for possibly the next 2 weeks.

**Comment by mrakitin** [comment]
➤ John Sinsheimer commented:

After a lot more puzzles early this week on this device, I'm now able to talk to the Scienta analyzer using the Jupyter notebook files provided by the company.  The IP address is 10.21.0.11 of the computer running the server that talks to the instrument and camera.  
The next steps are either of:
A.  Put these commands into a function callable by Bluesky to do stuff - although in any case it will take some effort to figure out what they want to do remotely.  The new software currently has no GUI and sparse documentation.
B.  Write an IOC and provide PVs - this is difficult since it uses http/1.1 protocol and/or websocket (EPICS IOC would require using asyn and C++ functions I think) or try caproto for interaction.

**Comment by mrakitin** [comment]
➤ Thomas Caswell commented:

Lets try caproto, if you can drive it from a jupyter notebook (I assume using requests or similar?) then we can drive it from caproto without too much trouble (ex see https://github.com/NSLS-II/caproto/blob/master/caproto/ioc_examples/currency_conversion.py).

**Comment by mrakitin** [comment]
➤ John Sinsheimer commented:

So the caproto documentation looks much improved since I last looked or remember!  Quick question though - if I want to run a caproto IOC on xf21id1-ioc1, should I make a new conda environment with Python 3.7+?  I couldn't find any existing environments on that machine.  Do you have any standard environments or methods to create environments that I could put on that machine?

**Comment by mrakitin** [comment]
➤ Dan Allan commented:

We don't deploy conda or any of our software on the -ioc machines. For development purposes, I suggest installing Python 3.6+ by whatever means is most convenient (conda, apt, compiling from source, etc.) and circling back once we're ready to operationalize it. FYI Python 3.7 works fine but 3.6 is sufficient.

**Comment by mrakitin** [comment]
➤ John Sinsheimer commented:

So I made my own conda environment and can run jupyter notebook either on the windows computer controlling the analyzer or xf21id1-ioc1, and can control the new software from either location.  We're now trying to figure out how to use the new software commands to do some typical configuration and measurement routines.  Although this is nontrivial, since the documentation is sparse and reading the API directly is possible, but not terribly helpful.  Planning to email Scienta with a bunch of questions this week on using their new software.

**Comment by mrakitin** [comment]
➤ Dan Allan commented:

Sounds good. Thanks for the update.

**Comment by mrakitin** [comment]
Good progress on that project!

**Comment by mrakitin** [comment]
➤ John Sinsheimer commented:

Another update:  Currently stuck on not knowing how to configure the element voltage for a measurement and how to subscribe to the camera image, due to lack of documentation.  Scienta says they don't have resources to spare right now as they're working on writing a GUI for the next 3 months.  Planning to beg them to answer these two questions soon and hope they can spare 5 minutes, because then we can actually do measurements with the new software.

Also python websockets module doesn't work from Jupyter notebook apparently (or Windows problem)?  

Also they had sent us the wrong camera and we just got the proper replacement yesterday and installed that.

**Comment by mrakitin** [comment]
Thanks for the update, [@John](https://3.basecamp-static.com/3733838/people/BAhpBKrZHwE=--6e9f88e30715ab4ad2cc07c3cd32c7ada0accbc0/avatar-20-x1?v=1). Hopefully the issues will be resolved, sorry to hear the customers are not the priority to the company. Is caproto already used anywhere in this project?

**Comment by mrakitin** [comment]
➤ John Sinsheimer commented:

Update:  We have the new software with GUI by Scienta.  Although there is almost no documentation, we've been able to get it working.  Scienta released new Jupyter notebook examples that actually work last Friday, and I'm now compiling those into basic routines like initialize, measure, export data, shut down, etc.  I've tested the new GUI from a remote machine and it mostly seems to work, but also need to test Python scripts.  Potentially in the next 2 weeks I can try putting these functions / user variables into a caproto IOC.  I'm anticipating some difficulties with handling events or waiting appropriately for functions to finish, which I have less experience with.

**Comment by mrakitin** [comment]
➤ Dan Allan commented:

Seems like "knowing when something is finished" is 80% of the work of interacting with hardware. I have some thoughts -- please reach out once you've got a sketch of Python scripts working and we try to contribute a useful suggestion or two.

**Comment by mrakitin** [comment]
Duplicate of #6.
